File: src/dynamic programming/can sum/src/can_sum.py
def can_sum_basic(target, numbers):

    if target == 0:
        return True

    if target < 0:
        return False

    for i in numbers:
        reminder = target - i
        if can_sum_basic(reminder, numbers):
            return True

    return False


def can_sum_memo(target, numbers, memo=dict()):

    if target == 0:
        return True

    if target < 0:
        return False

    if target in memo:
        return memo[target]

    for i in numbers:
        reminder = target - i
        memo[target] = can_sum_memo(reminder, numbers, memo)
        if memo[target]:
            return True

    return False


def can_sum_table(target, numbers):

    table = [False] * (target + 1)
    table[0] = True

    for i in range(target):
        if table[i]:
            numbers = [number for number in numbers if i + number <= target]
            for number in numbers:
                if i + number > 0 and i + number <= target:
                    table[i + number] = True

    return table[-1]

File: src/dynamic programming/cpp/can sum/src/can_sum.cpp
#include "can_sum.h"
#include <functional>
#include <unordered_map>

bool canSumBasic(const unsigned int target,
                 const std::vector<unsigned int> &numbers) {

  if (target == 0)
    return true;

  for (const auto &num : numbers) {

    if (num > target)
      return false;

    auto remainder = target - num;

    if (canSumBasic(remainder, numbers))
      return true;
  }

  return false;
}

bool canSumMemo(const unsigned int target,
                const std::vector<unsigned int> &numbers) {

  std::function<bool(const unsigned int &, const std::vector<unsigned int> &,
                     std::unordered_map<unsigned int, bool> &)>
      _canSumMemo;
  _canSumMemo = [&](const unsigned int target,
                    const std::vector<unsigned int> &numbers,
                    std::unordered_map<unsigned int, bool> &memo) -> bool {
    if (target == 0)
      return true;

    if (memo.count(target))
      return memo[target];

    for (const auto &num : numbers) {

      if (num > target)
        return false;

      auto remainder = target - num;

      if (canSumBasic(remainder, numbers)) {
        memo[target] = true;
        return true;
      }
    }

    memo[target] = false;
    return false;
  };

  std::unordered_map<unsigned int, bool> memo;

  return _canSumMemo(target, numbers, memo);
}

bool canSumTable(const unsigned int target, std::vector<unsigned int> numbers) {

  std::vector<bool> table(target + 1, false);
  table[0] = true;

  for (int i = 0; i < target; i++) {

    if (table[i]) {

      auto it = numbers.begin();
      while (it != numbers.end()) {

        if (i + *it <= target)
          it++;

        else
          it = numbers.erase(it);
      }

      for (const auto &number : numbers) {
        if (i + number > 0 && i + number <= target)
          table[i + number] = true;
      }
    }
  }

  return table.back();
}
